sled-agent/config-reconciler/src/ledger.rs

```diff
@@ -182,6 +246,891 @@ struct LedgerTask<T> {
+                        self.log,
+                        "rejecting config request due to \
+                         out-of-date generation";
+                        "current-gen" => %omicron_sled_config.generation,
+                        "request-gen" => %new_config.generation,
+                    );
+                    return Err(LedgerNewConfigError::GenerationOutdated {
+                        current: omicron_sled_config.generation,
+                        requested: new_config.generation,
+                    });
+                } else if new_config.generation
+                    == omicron_sled_config.generation
+                {
+                    if new_config != omicron_sled_config {
+                        warn!(
+                            self.log,
+                            "requested config changed (with same generation)";
+                            "generation" => %new_config.generation,
+                        );
+                        return Err(
+                            LedgerNewConfigError::ConfigurationChanged {
+                                generation: new_config.generation,
+                            },
+                        );
+                    }
+                    info!(
+                        self.log, "configuration unchanged";
+                        "generation" => %new_config.generation,
+                    );
+
+                    // We now fall through to committing the ledger to disk.
+                    // This seems unnecessary: if the generation matches, we
+                    // already committed the ledger previously. But we don't
+                    // know that it's been written to both M.2s:
+                    //
+                    // * It's possible only one M.2 was available when we
+                    //   previously committed
+                    // * It's possible we had two M.2s, but one of the writes
+                    //   failed (`Ledger::commit()` returns `Ok(())` as long as
+                    //   at least one of the writes succeeds)
+                    //
+                    // We could probably address both of those and avoid
+                    // spurious writes, but it seems a little tricky and not
+                    // particularly important. Sled configs are currently on the
+                    // order of 25 KiB. If we get a request from each Nexus
+                    // every minute, that's ~75 KiB of usually-spurious writes
+                    // to both M.2 drives every minute. Rounding up to 100
+                    // KiB/minute (to give some room for the config to grow),
+                    // that's ~150 MiB / day, and the our drive endurance is 1
+                    // TiB / day (1 DWPD). We're using up something like 0.01%
+                    // of the drive endurance via these writes.
+                }
+            }
+        }
+
+        // Continue validating the incoming config. For now, the only other
```

Comment:
Nit, for readability, we could consider pulling this out into a helper fn acting on `new_config`

Reply:
Yeah, that's much nicer, thanks: 41863807c